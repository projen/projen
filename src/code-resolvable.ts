const CODE_RESOLVABLE_SYMBOL = Symbol.for('projen.CodeResolvable');

export interface IImportResolver {
  /**
   * Adds a named import from a module
   */  
  from(moduleName: string, importName: string, as?: string): ICodeResolvable;
}

/**
 * Interface for code that should be rendered as-is without quotes.
 */
export interface ICodeResolvable {
  /**
   * Renders the code as a string.
   * @returns The rendered code
   */
  render(): string;
  
  /**
   * Resolves imports needed by this code resolvable.
   * @param resolver - The IImportsResolver instance to resolve imports into
   */
  resolveImports?(resolver: IImportResolver): void;
}

/**
 * Abstract base class for code that should be rendered as-is without quotes.
 */
export abstract class CodeResolvable implements ICodeResolvable {
  /**
   * Checks if an object is a CodeResolvable instance.
   * @param obj - The object to check
   * @returns True if the object is CodeResolvable
   */
  public static isCodeResolvable(obj: any): obj is ICodeResolvable {
    return obj && typeof obj === 'object' && obj[CODE_RESOLVABLE_SYMBOL] === true;
  }

  /**
   * Creates a new CodeResolvable instance and marks it with the symbol.
   */
  public constructor() {
    // mark as CodeResolvable
    Object.defineProperty(this, CODE_RESOLVABLE_SYMBOL, { value: true });
  }

  /**
   * Renders the code as a string.
   * @returns The rendered code
   */
  public abstract render(): string;
}
