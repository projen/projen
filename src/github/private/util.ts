export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

export function context(value: string) {
  return `\${{ ${value} }}`;
}

// Checks if part of the file path is hidden
export function isHiddenPath(path: string) {
  return /(^|\/)\.[^\/\.]/g.test(path);
}

// Helper to assert a path is not hidden
export function ensureNotHiddenPath(value: string, name: string) {
  if (isHiddenPath(value)) {
    throw Error(`${name} cannot be a hidden path, got: ${value}`);
  }
}
