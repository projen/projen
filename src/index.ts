export * from "./ai-instructions";
export * from "./compare";
export * from "./component";
export * from "./dev-env";
export * from "./dependencies";
export * from "./docker-compose";
export * from "./dockerfile";
export * from "./file";
export * from "./gitattributes";
export * from "./gitpod";
export * from "./runner-options";
export * from "./ignore-file";
export * from "./ini";
export * from "./json";
export * from "./json-patch";
export * from "./logger";
export * from "./license";
export * from "./makefile";
export * from "./object-file";
export * from "./option-hints";
export * from "./project";
export * from "./project-build";
export * from "./project-tree";
export * from "./projects";
export * from "./projenrc";
export * from "./projenrc-json";
export * from "./readme";
export * from "./renovatebot";
export * from "./sample-file";
export * from "./semver";
export * from "./source-code";
export * from "./task";
export * from "./tasks";
export * from "./task-model";
export * from "./task-runtime";
export * from "./testing";
export * from "./textfile";
export * from "./toml";
export * from "./version";
export * from "./yaml";
export * from "./xmlfile";

// export submodules
export * as web from "./web";
export * as github from "./github";
export * as gitlab from "./gitlab";
export * as circleci from "./circleci";
export * as vscode from "./vscode";
export * as java from "./java";
export * as python from "./python";
export * as typescript from "./typescript";
export * as javascript from "./javascript";
export * as release from "./release";
export * as awscdk from "./awscdk";
export * as cdk8s from "./cdk8s";
export * as cdktf from "./cdktf";
export * as cdk from "./cdk";
export * as build from "./build";
