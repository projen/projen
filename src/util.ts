import * as child_process from "child_process";
import {
  accessSync,
  chmodSync,
  constants as fs_constants,
  existsSync,
  mkdirSync,
  promises as fs,
  readFileSync,
  writeFileSync,
} from "fs";
import * as path from "path";
import * as Case from "case";
import * as logging from "./logging";

const MAX_BUFFER = 10 * 1024 * 1024;

/**
 * Executes a command with STDOUT > STDERR.
 */
export function exec(
  command: string,
  options: {
    cwd: string;
    env?: Record<string, string>;
    stdio?: child_process.StdioOptions;
  }
): void {
  logging.debug(`${command} (cwd: ${options.cwd})`);

  child_process.execSync(command, {
    stdio: options.stdio || ["inherit", 2, "pipe"], // "pipe" for STDERR means it appears in exceptions
    maxBuffer: MAX_BUFFER,
    cwd: options.cwd,
    env: options.env,
  });
}

/**
 * Executes command and returns STDOUT. If the command fails (non-zero), throws an error.
 */
export function execCapture(
  command: string,
  options: { cwd: string; modEnv?: Record<string, string> }
) {
  logging.debug(`${command} (cwd: ${options.cwd})`);

  return child_process.execSync(command, {
    stdio: ["inherit", "pipe", "pipe"], // "pipe" for STDERR means it appears in exceptions
    maxBuffer: MAX_BUFFER,
    cwd: options.cwd,
    env: {
      ...process.env,
      ...options.modEnv,
    },
  });
}

/**
 * Executes `command` and returns its value or undefined if the command failed.
 */
export function execOrUndefined(
  command: string,
  options: { cwd: string }
): string | undefined {
  logging.debug(`${command} (cwd: ${options.cwd})`);

  try {
    const value = child_process
      .execSync(command, {
        stdio: ["inherit", "pipe", "pipe"], // "pipe" for STDERR means it appears in exceptions
        maxBuffer: MAX_BUFFER,
        cwd: options.cwd,
      })
      .toString("utf-8")
      .trim();

    if (!value) {
      return undefined;
    } // an empty string is the same as undefined
    return value;
  } catch {
    return undefined;
  }
}

export interface WriteFileOptions {
  /**
   * Whether the generated file should be marked as executable.
   *
   * @default false
   */
  executable?: boolean;

  /**
   * Whether the generated file should be readonly.
   *
   * @default false
   */
  readonly?: boolean;
}

export function getFilePermissions(options: WriteFileOptions): string {
  const readonly = options.readonly ?? false;
  const executable = options.executable ?? false;
  if (readonly && executable) {
    return "544";
  } else if (readonly) {
    return "444";
  } else if (executable) {
    return "755";
  } else {
    return "644";
  }
}

export function writeFile(
  filePath: string,
  data: any,
  options: WriteFileOptions = {}
) {
  if (existsSync(filePath)) {
    chmodSync(filePath, "600");
  }

  mkdirSync(path.dirname(filePath), { recursive: true });
  writeFileSync(filePath, data);

  chmodSync(filePath, getFilePermissions(options));
}

/**
 * Decamelizes the keys of an object structure, recursing through child objects and arrays.
 * @experimental
 */
export interface DecamelizeRecursivelyOptions {
  /**
   * Max depth to recurse before erroring.
   * @default 10
   */
  maxDepth?: number;

  /**
   * Returns true when a key should be decamelized
   * @default - all keys are decamelized
   */
  shouldDecamelize?: (path: string[], value: any) => boolean;

  /**
   * Separator for decamelizing.
   * @default "_"
   */
  separator?: string;

  /**
   * Current path.
   * @internal
   */
  path?: string[];
}

export function decamelizeKeysRecursively(
  input: any,
  opt?: DecamelizeRecursivelyOptions
): any {
  const shouldAlwaysDecamelize = () => true;
  const shouldDecamelize = opt?.shouldDecamelize ?? shouldAlwaysDecamelize;
  const separator = opt?.separator ?? "_";
  const path_ = opt?.path ?? [];
  const maxDepth = opt?.maxDepth ?? 10;

  if (path_.length > maxDepth) {
    throw new Error(
      "Decamelled too deeply - check that the input has no circular references"
    );
  }

  if (Array.isArray(input)) {
    return input.map((k, i) =>
      decamelizeKeysRecursively(k, {
        ...opt,
        path: [...path_, i.toString()],
      })
    );
  }

  if (typeof input === "object" && input !== null) {
    const mappedObject: Record<string, any> = {};
    for (const [key, value] of Object.entries(input)) {
      const transformedKey = shouldDecamelize([...path_, key], value)
        ? decamelize(key, separator)
        : key;

      mappedObject[transformedKey] = decamelizeKeysRecursively(value, {
        ...opt,
        path: [...path_, key],
      });
    }

    return mappedObject;
  }

  return input;
}

/**
 * Returns false if value is unset or a falsey value, and true otherwise.
 * @param value an environment variable
 */
export function isTruthy(value: string | undefined): boolean {
  return !(
    value === undefined ||
    ["null", "undefined", "0", "false", ""].includes(value.toLocaleLowerCase())
  );
}

/**
 * Type of a map mapping strings to some arbitrary type
 */
export type Obj<T> = { [key: string]: T };

/**
 * Return whether the given value is an object
 *
 * Even though arrays and instances of classes technically are objects, we
 * usually want to treat them differently, so we return false in those cases.
 */
export function isObject(x: any): x is Obj<any> {
  return (
    x !== null &&
    typeof x === "object" &&
    !Array.isArray(x) &&
    x.constructor.name === "Object"
  );
}

/**
 * Recursively merge objects together
 *
 * The leftmost object is mutated and returned. Arrays are not merged
 * but overwritten just like scalars.
 *
 * If an object is merged into a non-object, the non-object is lost.
 *
 * `undefined`s will cause a value to be deleted if destructive is enabled.
 */
export function deepMerge(
  objects: Array<Obj<any> | undefined>,
  destructive: boolean = false
) {
  function mergeOne(target: Obj<any>, source: Obj<any>) {
    for (const key of Object.keys(source)) {
      const value = source[key];

      if (isObject(value)) {
        // if the value at the target is not an object, override it with an
        // object so we can continue the recursion
        if (typeof target[key] !== "object") {
          target[key] = value;
        }

        if ("__$APPEND" in value && Array.isArray(value.__$APPEND)) {
          if (Array.isArray(target[key])) {
            target[key].push(...value.__$APPEND);
          } else {
            target[key] = value.__$APPEND;
          }
        }

        mergeOne(target[key], value);

        // if the result of the merge is an empty object, it's because the
        // eventual value we assigned is `undefined`, and there are no
        // sibling concrete values alongside, so we can delete this tree.
        const output = target[key];
        if (
          typeof output === "object" &&
          Object.keys(output).length === 0 &&
          destructive
        ) {
          delete target[key];
        }
      } else if (value === undefined && destructive) {
        delete target[key];
      } else if (typeof value !== "undefined") {
        target[key] = value;
      }
    }
  }

  const others = objects.filter((x) => x != null) as Array<Obj<any>>;

  if (others.length === 0) {
    return {};
  }
  const into = others.splice(0, 1)[0];

  others.forEach((other) => mergeOne(into, other));
  return into;
}

/*
 * Deduplicate values in a list, returning a new array.
 * @param array list of values
 */
export function dedupArray<T>(array: T[]): T[] {
  return array.filter((val, idx) => array.indexOf(val) === idx);
}

/**
 * Returns a sorted version of `x` or `undefined` if it is an empty array or object.
 */
export function sorted<T>(x: T) {
  if (x == null) {
    return undefined;
  }
  if (Array.isArray(x)) {
    if (x.length === 0) {
      return undefined;
    }
    return (x as unknown[]).sort();
  } else if (typeof x === "object") {
    if (Object.keys(x).length === 0) {
      return undefined;
    }
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(x).sort(([l], [r]) =>
      l.localeCompare(r)
    )) {
      result[key] = value;
    }
    return result as T;
  } else {
    return x;
  }
}

export function formatAsPythonModule(name: string) {
  return name.replace(/-/g, "_").replace(/\./g, "_");
}

/**
 * Extract git version number from command line
 *
 * @param gitVersionOutput the output from `git version` CLI
 * @returns the version of git
 */
export function getGitVersion(gitVersionOutput: string) {
  const match = gitVersionOutput.match(/\d+.\d+.\d+/);
  if (!match) {
    throw new Error("Unable to retrieve git version");
  }

  return match[0];
}

export function kebabCaseKeys<T = unknown>(obj: T, recursive = true): T {
  if (typeof obj !== "object" || obj == null) {
    return obj;
  }

  if (Array.isArray(obj)) {
    if (recursive) {
      obj = obj.map((v) => kebabCaseKeys(v, recursive)) as any;
    }
    return obj;
  }

  const result: Record<string, unknown> = {};
  for (let [k, v] of Object.entries(obj)) {
    if (recursive) {
      v = kebabCaseKeys(v, recursive);
    }
    result[decamelize(k).replace(/_/gm, "-")] = v;
  }
  return result as any;
}

export function snakeCaseKeys<T = unknown>(
  obj: T,
  recursive = true,
  exclusiveForRecordKeys: string[] = []
): T {
  if (typeof obj !== "object" || obj == null) {
    return obj;
  }

  if (Array.isArray(obj)) {
    if (recursive) {
      obj = obj.map((v) =>
        snakeCaseKeys(v, recursive, exclusiveForRecordKeys)
      ) as any;
    }
    return obj;
  }

  const result: Record<string, unknown> = {};
  for (let [k, v] of Object.entries(obj)) {
    if (recursive) {
      v = snakeCaseKeys(v, recursive, exclusiveForRecordKeys);
    }
    const modifiedKey =
      exclusiveForRecordKeys.length == 0 || exclusiveForRecordKeys.includes(k)
        ? Case.snake(k)
        : k;
    result[modifiedKey] = v;
  }
  return result as any;
}

export async function tryReadFile(file: string) {
  if (!existsSync(file)) {
    return "";
  }

  return fs.readFile(file, "utf-8");
}

export function tryReadFileSync(file: string) {
  if (!existsSync(file)) {
    return undefined;
  }

  return readFileSync(file, "utf-8");
}

export function isWritable(file: string) {
  try {
    accessSync(file, fs_constants.W_OK);
    return true;
  } catch {
    return false;
  }
}

/**
 * Asserts that the file should be executable. Always returns true on Windows.
 *
 * In Windows, the executable attribute is stored in the system setting PATHEXT, not in each file. Then, checking for executability is equivalent to checking for existence. To bypass checking for executability, we always return true on Windows.
 *
 * @param filePath The path to the file
 * @param shouldBeExecutable Whether the file should be executable
 * @returns true if `filePath` executable attribute matches `shouldBeExecutable` or if the platform is Windows, false otherwise
 */
export function assertExecutablePermissions(
  filePath: string,
  shouldBeExecutable: boolean
): boolean {
  const isWindows = process.platform === "win32";
  if (isWindows) {
    return true;
  }

  const prevExecutable = isExecutable(filePath);

  return prevExecutable === shouldBeExecutable;
}

export function isExecutable(file: string) {
  try {
    accessSync(file, fs_constants.X_OK);

    return true;
  } catch (e) {
    return false;
  }
}

function decamelize(s: string, sep: string = "_") {
  if (Case.of(s) === "camel") {
    return Case.lower(s, sep);
  } else {
    return s;
  }
}

export function getNodeMajorVersion(): number | undefined {
  const match = process.version.match(/(\d+)\.(\d+)\.(\d+)/);
  if (match) {
    const [major] = match.slice(1).map((x) => parseInt(x));
    return major;
  }
  return undefined;
}

export function anySelected(options: (boolean | undefined)[]): boolean {
  return options.some((opt) => opt);
}

export function multipleSelected(options: (boolean | undefined)[]): boolean {
  return options.filter((opt) => opt).length > 1;
}

/**
 * Checks if a path is a FS root
 *
 * Optional uses a provided OS specific path implementation,
 * defaults to use the implementation for the current OS.
 *
 * @internal
 */
export function isRoot(dir: string, osPathLib: typeof path = path): boolean {
  const parent = osPathLib.dirname(dir);
  return parent === dir;
}

/**
 * Run up project tree to find a file or directory
 *
 * @param lookFor the file or directory to look for
 * @param cwd current working directory, must be an absolute path
 * @returns path to the file or directory we are looking for, undefined if not found
 */
export function findUp(
  lookFor: string,
  cwd: string = process.cwd()
): string | undefined {
  if (existsSync(path.join(cwd, lookFor))) {
    return cwd;
  }

  if (isRoot(cwd)) {
    // This is a root
    return undefined;
  }
  return findUp(lookFor, path.dirname(cwd));
}

/**
 * Normalizes a path that is going to be persisted to have a cross platform representation.
 *
 * Normalized paths can be persisted and doesn't need to be modified when the platform changes.
 * `normalizePersistedPath` takes care of platform-specific properties like directory separator.
 * It uses `path.posix.sep` that is supported both in Windows and Unix platforms.
 *
 *
 * @param p the path to be normalized
 * @returns the normalized path
 */
export function normalizePersistedPath(p: string) {
  return p.replace(/\\/g, path.posix.sep);
}
