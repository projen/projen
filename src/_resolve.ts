import { types } from "util";
import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // Special resolution for few JavaScript built-in types
  // that by default would be stringified as empty objects ('{}')
  // as they are missing a `toJSON` implementation.
  switch (true) {
    case types.isRegExp(value):
      if (value.flags) {
        throw new Error(
          "RegExp with flags should be explicitly converted to a string"
        );
      }
      return value.source;

    case types.isSet(value):
      return resolve(Array.from(value), options);

    case types.isMap(value):
      return resolve(Object.fromEntries(value), options);

    case types.isBoxedPrimitive(value):
      return resolve(value.valueOf(), options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
