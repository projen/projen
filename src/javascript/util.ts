import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve, posix } from "path";
import * as semver from "semver";
import { NodePackage } from "./node-package";
import { Project } from "../project";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(posix.sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return posix.join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given dependency version is installed
 *
 * This can be used to test for the presence of certain versions of devDependencies,
 * and do something dependency-specific in certain Components. For example, test for
 * a version of Jest and generate different configs based on the Jest version.
 *
 * NOTE: The implementation of this function currently is currently
 * approximate: to do it correctly, we would need a separate implementation
 * for every package manager, to query its installed version (either that, or we
 * would code to query `package-lock.json`, `yarn.lock`, etc...).
 *
 * Instead, we will look at `package.json`, and assume that the versions
 * picked by the package manager match ~that. This will work well enough for
 * major version checks, but may fail for point versions.
 *
 * What we SHOULD do is: `actualVersion ∈ checkRange`.
 *
 * What we do instead is a slightly more sophisticated version of
 * `requestedRange ∩ checkRange != ∅`. This will always give a correct result if
 * `requestedRange ⊆ checkRange`, but may give false positives when `checkRange
 * ⊆ requestedRange`.
 *
 * May return `undefined` if the question cannot be answered. These include the
 * following cases:
 *
 *   - The dependency is requested via local file dependencies (`file://...`)
 *   - The dependency uses an other type of URL, such as a GitHub URL
 *   - The dependency is not found in the `package.json`, such as when
 *     bootstrapping from an external projen package, and the `package.json`
 *     file only has that one external package as a dependency
 *
 * Otherwise it will return `true` if the installed version is probably in the
 * requested range, and `false` if it is probably not.
 *
 * This API may eventually be added to the public projen API, but only after
 * we implement exact version checking.
 *
 * @param dependencyName The name of the dependency
 * @param checkRange A particular version, or range of versions.
 */
export function hasDependencyVersion(
  project: Project,
  dependencyName: string,
  checkRange: string
): boolean | undefined {
  const file = NodePackage.of(project)?.file;
  if (!file) {
    return undefined;
  }

  if (!existsSync(file.absolutePath)) {
    return undefined;
  }

  const pj = JSON.parse(readFileSync(file.absolutePath, "utf-8"));

  // Technicaly, we should be intersecting all ranges to come up with the most narrow dependency
  // range, but `semver` doesn't allow doing that and we don't want to add a dependency on `semver-intersect`.
  //
  // Let's take the first dependency declaration we find, and assume that people
  // set up their `package.json` correctly.
  let requestedRange: string | undefined;
  for (const key of ["dependencies", "devDependencies", "peerDependencies"]) {
    const deps = pj[key] ?? {};
    let requestedVersion = deps[dependencyName];
    if (requestedVersion) {
      // If this is not a valid range, it could be 'file:dep.tgz', or a GitHub URL. No way to know what
      // version we're getting, bail out.
      if (!semver.validRange(requestedVersion)) {
        return undefined;
      }
      requestedRange = requestedVersion;
      break;
    }
  }

  // If the dependency is not found in the `package.json`, we can't answer the question (yet).
  // This means that the dependency hasn't been added yet, which means we know (upstream) what we're going to request,
  // or we're going to ask for '*' and we'll get the latest version.
  if (!requestedRange) {
    return undefined;
  }

  return installedVersionProbablyMatches(requestedRange, checkRange);
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
