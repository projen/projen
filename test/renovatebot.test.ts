import { synthSnapshot, TestProject } from "./util";
import { DependencyType } from "../src";

describe("renovatebot", () => {
  test("renovatebot: true creates renovatebot configuration", () => {
    // GIVEN
    const p = new TestProject({
      renovatebot: true,
      renovatebotOptions: {
        labels: ["renotate", "dependencies"],
      },
    });

    // WHEN
    const snapshot = synthSnapshot(p);

    // THEN
    expect(snapshot["renovate.json5"]).toMatchSnapshot();
  });

  test("renovatebot: will ignore deps added later", () => {
    // GIVEN
    const p = new TestProject({
      renovatebot: true,
      renovatebotOptions: {
        labels: ["renotate", "dependencies"],
      },
    });
    p.deps.addDependency("test@1.0.0", DependencyType.RUNTIME);

    // WHEN
    const snapshot = synthSnapshot(p);

    // THEN
    expect(snapshot["renovate.json5"]).toHaveProperty("ignoreDeps", [
      "test",
      "projen",
      "amannn/action-semantic-pull-request",
    ]);
  });

  test("renovatebot: will ignore workflow actions", () => {
    // GIVEN
    const p = new TestProject({
      renovatebot: true,
      renovatebotOptions: {
        labels: ["renotate", "dependencies"],
      },
      github: true,
      githubOptions: {
        workflows: true,
      },
    });

    // WHEN
    const snapshot = synthSnapshot(p);

    // THEN
    expect(snapshot["renovate.json5"]).toHaveProperty("ignoreDeps", [
      "projen",
      "amannn/action-semantic-pull-request",
    ]);
  });

  test("renovatebot: will ignore reusable workflow versions", () => {
    // GIVEN
    const p = new TestProject({
      renovatebot: true,
      renovatebotOptions: {
        labels: ["renotate", "dependencies"],
      },
      github: true,
      githubOptions: {
        workflows: true,
      },
    });

    const workflow = p.github?.addWorkflow("self-added");
    workflow?.addJob("actions", {
      permissions: {},
      uses: "own-org/reusable-workflows/.github/workflows/testing.yaml@v1",
    });

    // WHEN
    const snapshot = synthSnapshot(p);

    // THEN
    expect(snapshot["renovate.json5"]).toHaveProperty("ignoreDeps", [
      "projen",
      "amannn/action-semantic-pull-request",
      "own-org/reusable-workflows/.github/workflows/testing.yaml",
    ]);
  });

  test("renovatebot: will ignore workflow actions created manually", () => {
    // GIVEN
    const p = new TestProject({
      renovatebot: true,
      renovatebotOptions: {
        labels: ["renotate", "dependencies"],
      },
      github: true,
      githubOptions: {
        workflows: true,
      },
    });

    const workflow = p.github?.addWorkflow("self-added");
    workflow?.addJob("actions", {
      permissions: {},
      steps: [
        {
          name: "checkout",
          uses: "actions/checkout@v4",
        },
        {
          name: "cache",
          uses: "actions/cache@v4",
        },
      ],
    });
    // WHEN
    const snapshot = synthSnapshot(p);

    // THEN
    expect(snapshot["renovate.json5"]).toHaveProperty("ignoreDeps", [
      "projen",
      "amannn/action-semantic-pull-request",
      "actions/checkout",
      "actions/cache",
    ]);
  });

  test("renovatebot: override renovatebot configuration", () => {
    // GIVEN
    const overrideConfig = {
      prHourlyLimit: 10,
      schedule: [],
      packageRules: [],
      labels: [],
      ignoreDeps: [],
      extends: ["customExtend"],
    };
    const p = new TestProject({
      renovatebot: true,
      renovatebotOptions: {
        marker: false,
        labels: ["renotate", "dependencies"],
        overrideConfig: overrideConfig,
      },
    });

    // WHEN
    const snapshot = synthSnapshot(p)["renovate.json5"];

    // THEN
    expect(snapshot).toMatchSnapshot();
    expect(snapshot).toStrictEqual(overrideConfig);
  });

  test("renovatebot: can use file escape hatch", () => {
    // GIVEN
    const p = new TestProject({
      renovatebot: true,
      renovatebotOptions: {
        labels: ["renotate", "dependencies"],
      },
    });

    // THEN
    expect(p.tryFindObjectFile("renovate.json5")).toBeDefined();
  });
});
