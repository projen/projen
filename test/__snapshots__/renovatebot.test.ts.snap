// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renovatebot renovatebot: override renovatebot configuration 1`] = `
{
  "extends": [
    "customExtend",
  ],
  "ignoreDeps": [],
  "labels": [],
  "packageRules": [],
  "prHourlyLimit": 10,
  "schedule": [],
}
`;

exports[`renovatebot renovatebot: true creates renovatebot configuration 1`] = `
{
  "extends": [
    ":preserveSemverRanges",
    "config:recommended",
    "group:allNonMajor",
    "group:recommended",
    "group:monorepos",
  ],
  "ignoreDeps": [
    "projen",
    "amannn/action-semantic-pull-request",
  ],
  "labels": [
    "renotate",
    "dependencies",
  ],
  "packageRules": [
    {
      "groupName": "devDependencies (non-major)",
      "matchDepTypes": [
        "devDependencies",
      ],
      "matchUpdateTypes": [
        "patch",
        "minor",
      ],
    },
  ],
  "schedule": [
    "at any time",
  ],
}
`;
